### ISSUE TITLE
Create a eks-machinepool flavor

### ISSUE THREAD

**Issue by richardcase** [kubernetes-sigs/cluster-api-provider-aws#2029]
/kind feature
/area provider/eks
/help
/good-first-issue

**Describe the solution you'd like**
We should create a new template for a `eks-machinepool` flavor. This is so users can use clusterctl to create an EKS cluster that uses a machine pool. For example:

```bash
clusterctl config cluster my-cluster --kubernetes-version v1.16.8 --flavor eks-machinepool > my-cluster.yaml
```

This will invlove creating the template file and also updating the makefile to copy the template to the out folder and updating the release docs. 

**Anything else you would like to add:**
This is a follow up from #1863 and is the umanaged version of #2024 

**Environment:**

- Cluster-api-provider-aws version: 0.6.0

**Comment by kkeshavamurthy** [comment]
I'll give this a try. 
/assign

**Comment by richardcase** [comment]
@kkeshavamurthy - reach out here or on slack if you need anything and thanks for working on this.

**Comment by kkeshavamurthy** [comment]
@richardcase, Thanks I definitely will. Initially I thought I was supposed to do the flavor for eks-machinepool (Managed), hadn't noticed that it was already done. I'm not sure what unmanaged eks machinepool flavor looks like. I'll look into it a bit more. Any pointers would be very useful.

**Comment by richardcase** [comment]
The API definition for AWSMachinePool is here: https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/exp/api/v1alpha3/awsmachinepool_types.go#L145

I think the falvor should have the minimum set of fields.

**Comment by sedefsavas** [comment]
/assign @richardcase 
For triage.

**Comment by richardcase** [comment]
This is a nice to have.

/milestone v0.7.x
/priority backlog
/unassign

**Comment by vibhorrawat** [comment]
Hi Folks, I'm a new contributor and willing to work on this _good first issue_. However, from the comments it looks like this was picked and then nothing has happened. Hence, just to confirm is this still a valid/on-going issue? If yes, may I pick this up?

**Comment by richardcase** [comment]
Hey @vibhorrawat  - thanks for considering contributing. This is still a valid issue to work on. If you want to work on it feel free  to assign it yourself and mark it active (see [this](https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2224#issuecomment-925660001) comment for an example of how to do that). If you have any questions along the way feel free to ping me here or in the slack channel.

**Comment by vibhorrawat** [comment]
@richardcase Great! Thank you!
/assign

**Comment by vibhorrawat** [comment]
@richardcase Here is my understanding of the problem statement please feel free to correct me if I'm wrong here.

- Create a new template (a yaml file) for `eks-machinepool`. For this, when I started looking at the [templates](https://github.com/kubernetes-sigs/cluster-api-provider-aws/tree/main/templates) folder, we have few templates like `template-eks` and `template-machinepool`. And now we need a new one which will be a combination of both? 
- Second, _updating the makefile to copy the template to the out folder._ Are you referring to this [line](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/b2ed789aa803401cbbcadd07cf9ecead97ffc52c/Makefile#L498) or is there something else required here. Would it be possible for you to share the pointers.
- Third, release docs are [here](https://github.com/vibhorrawat/cluster-api-provider-aws/blob/main/docs/book/src/development/releasing.md) Kindly confirm.

Since, I'm a newbie to k8s, I will try to setup the development environment locally by following the [guide](https://cluster-api-aws.sigs.k8s.io/development/development.html).

Please share your thoughts.

**Comment by richardcase** [comment]
Hi @vibhorrawat 

You are understanding is correct :+1: 

With `update the makefile to copye the template` this is already done. As well as updating the release notes documentation i'd also add a note to the [machine pool docs](https://github.com/vibhorrawat/cluster-api-provider-aws/blob/main/docs/book/src/topics/machinepools.md) to mention that there is a EKS with machine pool template as well.

**Comment by vibhorrawat** [comment]
@richardcase question, does the [Set up AWS Environment](https://cluster-api-aws.sigs.k8s.io/development/development.html#setup-aws-environment) requires a separate aws account provided by CNCF or I can configure mine? I did not understand the suggested steps in documentation. Apologies for that. May be I lack the context of how `clusterawsadm` deploys the infrastructure on AWS. 

Besides, for the template could you please help me with what `kind` should go in? Referring these two

- `cluster-template-machinepool.yaml`
- `cluster-template-eks.yaml`

For instance-
The `Cluster` should have infrastructureRef as `AWSCluster` and controlPlaneRef as `KubeadmControlPlane` instead of `AWSManagedControlPlane`? or 
And we will use `EKSConfigTemplate` for `bootstrap` in `MachinePool`

Please suggest. We can connect over slack if you around. Thank you

**Comment by Jacobious52** [comment]
@richardcase - Thanks for pointing me in the right direction, I found https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/7c7ebfe7e2256a092caa8df1481a838feee35836/bootstrap/eks/controllers/eksconfig_controller.go#L309 and that the reason it wasn't reconciling for me was I referenced an `EKSConfigTemplate` like in the `MachineDeployment` flavour, instead of a concrete `EKSConfig`. When I tried creating an `EKSConfig` directly, I didn't update the `MachinePool`'s reference to the new object. This makes more sense now given the eksconfig_controller is in charge of adding the MachinePool as the owner so it gets properly reconciled.

Correctly referencing the `EKSConfig` works for me now 👍 I guess adding tests will still be beneficial anyway.

Having this flavour provided in the templates should help others to avoid that confusion. Thanks again!

---

@vibhorrawat - For reference if it helps for your template my working bootstrap for `MachinePool` looks like below without the templating vars.
```yaml
spec:
  bootstrap:
    configRef:
      apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
      kind: EKSConfig
      name: mp-eks-config
---
apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
kind: EKSConfig
metadata:
  name: mp-eks-config
```
And for the `Cluster` refs, both reference the `AWSManagedControlPlane`
```yaml
spec:
  controlPlaneRef:
    apiVersion: controlplane.cluster.x-k8s.io/v1beta1
    kind: AWSManagedControlPlane
    name: mp-control-plane
  infrastructureRef:
    apiVersion: controlplane.cluster.x-k8s.io/v1beta1
    kind: AWSManagedControlPlane
    name: mp-control-plane
```

**Comment by vibhorrawat** [comment]
@richardcase Hey there, Latest on this issue- I'm ready with the required template (added the aforementioned suggestions); however, I'm struggling to test that locally. 

Here is what I have done till date-

- I have followed the developer [guide](https://cluster-api-aws.sigs.k8s.io/development/tilt-setup.html#run-tilt) to build the `kind cluster` locally using `tilt`. Thanks to @sedefsavas I was unblocked and was enabled to execute `tilt up` command.
- Next step was to validate if pods are up & running on the `kind cluster`. This is where the problem is, I'm unable to move forward. When I execute the following command `kubectl get pods -A.` there is no response or error 😞 
- After this, the important step is to spin up [CAPA managed workload cluster](https://cluster-api-aws.sigs.k8s.io/development/tilt-setup.html#create-workload-clusters) using required (in my case, using newly created) template.

Your thoughts, any pointers to proceed?

Thank you!

**Comment by vibhorrawat** [comment]
Thank you @sedefsavas. I will give another try for it.

**Comment by richardcase** [comment]
> Thank you @sedefsavas. I will give another try for it.

@vibhorrawat - feel free to ping me on slack as well if you need help.

**Comment by Ankitasw** [comment]
Hey @vibhorrawat are you still working on this issue?

**Comment by Ankitasw** [comment]
@vibhorrawat yes, please share your observation, will try to help.

**Comment by sedefsavas** [comment]
@vibhorrawat 
There is an example here https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3094/files for AWSManagedMachinePool with LaunchTemplates.
Might be useful as a reference.

**Comment by vibhorrawat** [comment]
@Ankitasw Hi, here is what I have observed during the test run locally-

It looks like the test either failed or running into a loop. I do see a VPC, few subnets, security groups and NAT gateways getting creating in the AWS. However, it is not moving ahead after that.  There are few warnings but no errors in Tilt console.
I'm seeing following error message in `capi_controller`.
```
I1123 17:12:24.185244       9 machinepool_controller_phases.go:271] controller/machinepool "msg"="Infrastructure provider is not ready, requeuing" "cluster"="capi-test-eks-machine-flavor" "name"="capi-test-eks-machine-flavor-mp-0" "namespace"="default" "reconciler group"="cluster.x-k8s.io" "reconciler kind"="MachinePool" 
E1123 17:12:24.186380       9 controller.go:317] controller/machinepool "msg"="Reconciler error" "error"="could not find bootstrap.cluster.x-k8s.io/v1beta1, Kind=EKSConfigTemplate \"capi-test-eks-machine-flavor-mp-0\" for MachinePool \"capi-test-eks-machine-flavor-mp-0\" in namespace \"default\", requeuing: failed to retrieve EKSConfigTemplate external object \"default\"/\"capi-test-eks-machine-flavor-mp-0\": eksconfigtemplates.bootstrap.cluster.x-k8s.io \"capi-test-eks-machine-flavor-mp-0\" not found" "name"="capi-test-eks-machine-flavor-mp-0" "namespace"="default" "reconciler group"="cluster.x-k8s.io" "reconciler kind"="MachinePool" 
```
In `capa_controller` I see following-
```
I1123 17:09:30.024059      10 awsmachinepool_controller.go:219]  "msg"="Bootstrap data secret reference is not yet available"  
I1123 17:09:35.214427      10 subnets.go:47]  "msg"="Reconciling subnets"  
I1123 17:09:38.309111      10 roles.go:67]  "msg"="using eks control plane role"  "role-name"="eks-controlplane.cluster-api-provider-aws.sigs.k8s.io"
I1123 17:09:41.771436      10 securitygroup.go:32]  "msg"="Reconciling EKS security groups"  "cluster-name"="default_capi-test-eks-machine-flavor-control-plane"
I1123 17:09:42.663540      10 cluster.go:496]  "msg"="reconciling encryption configuration"  
I1123 17:09:42.674559      10 addons.go:34]  "msg"="Reconciling EKS addons"  
I1123 17:09:43.096650      10 addons.go:105]  "msg"="no eks addons installed in cluster"  "cluster"="default_capi-test-eks-machine-flavor-control-plane"
I1123 17:09:43.097000      10 addons.go:57]  "msg"="no addons installed and no addons to install, no action needed"  
I1123 17:09:43.097774      10 identity_provider.go:32]  "msg"="reconciling oidc identity provider"  
I1123 17:09:43.098875      10 cni.go:45]  "msg"="Reconciling aws-node DaemonSet in cluster"  "cluster-name"="capi-test-eks-machine-flavor" "cluster-namespace"="default"
I1123 17:09:43.099584      10 reconcile.go:37]  "msg"="Reconciling aws-iam-authenticator configuration"  "cluster-name"="capi-test-eks-machine-flavor"
I1123 17:09:46.746880      10 reconcile.go:96]  "msg"="Reconciled aws-iam-authenticator configuration"  "cluster-name"="default_capi-test-eks-machine-flavor-control-plane"
```
Few more events I saw-
```
cabpk_contro… │ [K8s EVENT: Pod capi-kubeadm-bootstrap-controller-manager-7bb97dcd98-rqjtz (ns: capi-kubeadm-bootstrap-system)] Liveness probe failed: Get "http://10.244.0.8:9440/healthz": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
kcp_controll… │ [K8s EVENT: Pod capi-kubeadm-control-plane-controller-manager-7bf7fc555-xpnpp (ns: capi-kubeadm-control-plane-system)] Liveness probe failed: Get "http://10.244.0.6:9440/healthz": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
capi_control… │ I1123 17:12:24.185244       9 machinepool_controller_phases.go:271] controller/machinepool "msg"="Infrastructure provider is not ready, requeuing" "cluster"="capi-test-eks-machine-flavor" "name"="capi-test-eks-machine-flavor-mp-0" "namespace"="default" "reconciler group"="cluster.x-k8s.io" "reconciler kind"="MachinePool" 
```

Kindly suggest! I can share the template too if required.

Thanks!

**Comment by Ankitasw** [comment]
Hey @vibhorrawat, while the test runs, can you do `kubectl describe clusters` and see at which stage it's stuck?

**Comment by richardcase** [comment]
It shoulds like the `EKSConfig` isn't there. 

It might also be worth upping the verbosity level of the controller.

**Comment by vibhorrawat** [comment]
@Ankitasw Here is the output of the `kubectl describe clusters`-
```
Name:         capi-test-eks-machine-flavor
Namespace:    default
Labels:       <none>
Annotations:  <none>
API Version:  cluster.x-k8s.io/v1beta1
Kind:         Cluster
Metadata:
  Creation Timestamp:  2021-11-23T14:34:28Z
  Finalizers:
    cluster.cluster.x-k8s.io
  Generation:  2
  Managed Fields:
    API Version:  cluster.x-k8s.io/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:clusterNetwork:
          .:
          f:pods:
            .:
            f:cidrBlocks:
        f:controlPlaneRef:
          .:
          f:apiVersion:
          f:kind:
          f:name:
        f:infrastructureRef:
          .:
          f:apiVersion:
          f:kind:
          f:name:
    Manager:      kubectl-client-side-apply
    Operation:    Update
    Time:         2021-11-23T14:34:28Z
    API Version:  cluster.x-k8s.io/v1beta1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .:
          v:"cluster.cluster.x-k8s.io":
      f:spec:
        f:controlPlaneEndpoint:
          f:host:
          f:port:
      f:status:
        .:
        f:conditions:
        f:controlPlaneReady:
        f:failureDomains:
          .:
          f:us-west-2a:
            .:
            f:controlPlane:
          f:us-west-2b:
            .:
            f:controlPlane:
          f:us-west-2c:
            .:
            f:controlPlane:
        f:infrastructureReady:
        f:observedGeneration:
        f:phase:
    Manager:         manager
    Operation:       Update
    Time:            2021-11-23T14:50:29Z
  Resource Version:  1182442
  UID:               38bcd390-15b1-4668-ae9b-e877ee1bb68b
Spec:
  Cluster Network:
    Pods:
      Cidr Blocks:
        192.168.0.0/16
  Control Plane Endpoint:
    Host:  <valid_host_url>
    Port:  443
  Control Plane Ref:
    API Version:  controlplane.cluster.x-k8s.io/v1beta1
    Kind:         AWSManagedControlPlane
    Name:         capi-test-eks-machine-flavor-control-plane
    Namespace:    default
  Infrastructure Ref:
    API Version:  controlplane.cluster.x-k8s.io/v1beta1
    Kind:         AWSManagedControlPlane
    Name:         capi-test-eks-machine-flavor-control-plane
    Namespace:    default
Status:
  Conditions:
    Last Transition Time:  2021-12-10T12:02:30Z
    Status:                True
    Type:                  Ready
    Last Transition Time:  2021-11-23T14:50:29Z
    Status:                True
    Type:                  ControlPlaneInitialized
    Last Transition Time:  2021-12-10T12:02:30Z
    Status:                True
    Type:                  ControlPlaneReady
    Last Transition Time:  2021-12-10T12:02:30Z
    Status:                True
    Type:                  InfrastructureReady
  Control Plane Ready:     true
  Failure Domains:
    us-west-2a:
      Control Plane:  true
    us-west-2b:
      Control Plane:  true
    us-west-2c:
      Control Plane:     true
  Infrastructure Ready:  true
  Observed Generation:   2
  Phase:                 Provisioned
Events:                  <none>

```
Kindly note I have masked the control plane's host url

Besides, as @richardcase suggest I do see following error in `capi_controller` logs.
```
E0315 17:29:41.453897       9 controller.go:317] controller/machinepool "msg"="Reconciler error" "error"="could not find bootstrap.cluster.x-k8s.io/v1beta1, Kind=EKSConfigTemplate \"capi-test-eks-machine-flavor-mp-0\" for MachinePool \"capi-test-eks-machine-flavor-mp-0\" in namespace \"default\", requeuing: failed to retrieve EKSConfigTemplate external object \"default\"/\"capi-test-eks-machine-flavor-mp-0\": eksconfigtemplates.bootstrap.cluster.x-k8s.io \"capi-test-eks-machine-flavor-mp-0\" not found" "name"="capi-test-eks-machine-flavor-mp-0" "namespace"="default" "reconciler group"="cluster.x-k8s.io" "reconciler kind"="MachinePool" 
```

Kindly suggest the how could I investigate this further with increased verbose logging. 
Thank you.

**Comment by richardcase** [comment]
@vibhorrawat - could you share the full santized set of resource kinds for the cluster?

**Comment by vibhorrawat** [comment]
@richardcase  - Sure thing. I'm pretty new to this stuff and terminologies so apologies for asking this question. 
Could you please help me with the steps to get the required information? Is it a command or a documentation I can follow. Thank you.

**Comment by richardcase** [comment]
Sorry @vibhorrawat - i should've said its just the yaml for your cluster that you kubectl applied

**Comment by vibhorrawat** [comment]
@richardcase - Sure. Here it is- `templates/cluster-template-eks-machinepool.yaml`

```
---
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  name: "${CLUSTER_NAME}"
spec:
  clusterNetwork:
    pods:
      cidrBlocks: ["192.168.0.0/16"]
  infrastructureRef:
    kind: AWSManagedControlPlane
    apiVersion: controlplane.cluster.x-k8s.io/v1beta1
    name: "${CLUSTER_NAME}-control-plane"
  controlPlaneRef:
    kind: AWSManagedControlPlane
    apiVersion: controlplane.cluster.x-k8s.io/v1beta1
    name: "${CLUSTER_NAME}-control-plane"
---
kind: AWSManagedControlPlane
apiVersion: controlplane.cluster.x-k8s.io/v1beta1
metadata:
  name: "${CLUSTER_NAME}-control-plane"
spec:
  region: "${AWS_REGION}"
  sshKeyName: "${AWS_SSH_KEY_NAME}"
  version: "${KUBERNETES_VERSION}"
---
apiVersion: cluster.x-k8s.io/v1beta1
kind: MachinePool
metadata:
  name: ${CLUSTER_NAME}-mp-0
  namespace: default
spec:
  clusterName: ${CLUSTER_NAME}
  replicas: ${WORKER_MACHINE_COUNT}
  template:
    spec:
      bootstrap:
        configRef:
          apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
          kind: EKSConfigTemplate
          name: ${CLUSTER_NAME}-mp-0
      clusterName: ${CLUSTER_NAME}
      infrastructureRef:
        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
        kind: AWSMachinePool
        name: ${CLUSTER_NAME}-mp-0
      version: ${KUBERNETES_VERSION}
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: AWSMachinePool
metadata:
  name: ${CLUSTER_NAME}-mp-0
  namespace: default
spec:
  minSize: 1
  maxSize: 10
  availabilityZones:
    - "${AWS_REGION}"
  awsLaunchTemplate:
    instanceType: "${AWS_CONTROL_PLANE_MACHINE_TYPE}"
    sshKeyName: "${AWS_SSH_KEY_NAME}"
---
apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
kind: EKSConfigTemplate
metadata:
  name: "${CLUSTER_NAME}-md-0"
spec:
  template: {}
```

**Comment by pydctw** [comment]
EKSConfigTemplate referenced in MachinePool doesn't match the name in EKSConfigTemplate definition and that is why you are getting `failed to retrieve EKSConfigTemplate` error.

MachinePool
```
apiVersion: cluster.x-k8s.io/v1beta1
kind: MachinePool
...
spec:
   ...
    spec:
      bootstrap:
        configRef:
          apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
          kind: EKSConfigTemplate
          name: ${CLUSTER_NAME}-mp-0 ⬅️
```

EKSConfigTemplate
```
apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
kind: EKSConfigTemplate
metadata:
  name: "${CLUSTER_NAME}-md-0" ⬅️
spec:
  template: {}
```

**Comment by vibhorrawat** [comment]
@pydctw Thank you for pointing out the issue. Kindly allow me to do the change, test the template again and update with the latest. Thank you.

**Comment by vibhorrawat** [comment]
Hi @pydctw  and @richardcase, I think it worked. No more `failed to retrieve EKSConfigTemplate` error messages in the `capi _controller`.
Here is what I'm seeing in the controller logs-

In `capi_controller`::
```
I0331 10:20:42.371254       9 machinepool_controller_phases.go:271] controller/machinepool "msg"="Infrastructure provider is not ready, requeuing" "cluster"="capi-test-eks-machine-flavor" "name"="capi-test-eks-machine-flavor-mp-0" "namespace"="default" "reconciler group"="cluster.x-k8s.io" "reconciler kind"="MachinePool" 
I0331 10:21:12.372734       9 machinepool_controller_phases.go:271] controller/machinepool "msg"="Infrastructure provider is not ready, requeuing" "cluster"="capi-test-eks-machine-flavor" "name"="capi-test-eks-machine-flavor-mp-0" "namespace"="default" "reconciler group"="cluster.x-k8s.io" "reconciler kind"="MachinePool" 
```
In `capa_controller`::
```
I0331 10:19:51.213477       8 securitygroup.go:32]  "msg"="Reconciling EKS security groups"  "cluster-name"="default_capi-test-eks-machine-flavor-control-plane"
I0331 10:19:51.946636       8 cluster.go:496]  "msg"="reconciling encryption configuration"  
I0331 10:19:51.946720       8 addons.go:34]  "msg"="Reconciling EKS addons"  
I0331 10:19:52.311191       8 addons.go:105]  "msg"="no eks addons installed in cluster"  "cluster"="default_capi-test-eks-machine-flavor-control-plane"
I0331 10:19:52.311524       8 addons.go:57]  "msg"="no addons installed and no addons to install, no action needed"  
I0331 10:19:52.311604       8 identity_provider.go:32]  "msg"="reconciling oidc identity provider"  
I0331 10:19:52.312190       8 cni.go:45]  "msg"="Reconciling aws-node DaemonSet in cluster"  "cluster-name"="capi-test-eks-machine-flavor" "cluster-namespace"="default"
I0331 10:19:52.312236       8 reconcile.go:37]  "msg"="Reconciling aws-iam-authenticator configuration"  "cluster-name"="capi-test-eks-machine-flavor"
I0331 10:19:55.120387       8 reconcile.go:96]  "msg"="Reconciled aws-iam-authenticator configuration"  "cluster-name"="default_capi-test-eks-machine-flavor-control-plane"
```

Question- How could I validate if all the required stuff is deployed using the template? Any command or documentation I could refer to? Kindly suggest.

**Comment by richardcase** [comment]
@vibhorrawat - do you see the ASG and EC2 instances created for this cluster / machine pool?

**Comment by vibhorrawat** [comment]
@richardcase Nope I do not see any ASGs or EC2 instances created for the cluster. All I can see a cluster created in EKS.

**Comment by anmolsachan** [comment]
@richardcase @vibhorrawat I tried using awsmachinepool with eks as well and it does not work. Both CAPI and CAPA controllers keep waiting:

CAPI:

`I0517 10:02:47.001148       1 machinepool_controller_phases.go:271] controller/machinepool "msg"="Infrastructure provider is not ready, requeuing" "cluster"="capi-anmol-eks-machinepool" "name"="capi-anmol-eks-machinepool-mp-0" "namespace"="aws-test" "reconciler group"="cluster.x-k8s.io" "reconciler kind"="MachinePool"`

CAPA:

`I0517 10:07:54.937724       1 awsmachinepool_controller.go:193] controller/awsmachinepool "msg"="Reconciling AWSMachinePool" "cluster"="capi-anmol-eks-machinepool" "machinePool"="capi-anmol-eks-machinepool-mp-0" "name"="capi-anmol-eks-machinepool-mp-0" "namespace"="aws-test" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="AWSMachinePool" 
I0517 10:07:54.938216       1 awsmachinepool_controller.go:220]  "msg"="Bootstrap data secret reference is not yet available"  `

Looks like there is a bug.

**Comment by pydctw** [comment]
@vibhorrawat, are you still working on it? If not, I would like to look into it to see if there is a real bug.

**Comment by vibhorrawat** [comment]
Hey @pydctw I would love to continue on this issue; however, as I have updated perviously I'm stuck and not sure how to progress. Let me know your thoughts. Thank you!

**Comment by richardcase** [comment]
@pydctw - did you want to look into this? Or shall i take a look to see if i can help out?

**Comment by pydctw** [comment]
I figured it out a few days ago and working on [adding e2e test](https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2865), which is really needed.
